### PR TITLE
Upgrade to recent FluentD and Ruby versions

### PR DESCRIPTION
**Pull request by markbreedlove**
This PR makes a number of changes so that the formula works with more recent o/s, Ruby, and FluentD versions. It also makes sure that FluentD runs as the configured user. (It was creating the user, but not ensuring that the daemon runs as that user.)

Updates:

* Use `rbenv` to install the same currently-updated version of Ruby, regardless of o/s.
* Run `fluentd` as the configured user, by default `fluentd`.
* Update Vagrant boxes to validate the formula on modern Debian, Ubuntu, and CentOS machines.
* Fix the syntax of FluentD config files to account for changes from version 0.x to version 1.x.
* Fix Salt states to account for changes in rendering behavior in either Salt or Jinja.
* Fix issues with service starts and restarts.

You should be able to bring up all three Vagrant VMs with `vagrant up` and the `testinfra` suites should pass.

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -82,11 +95,13 @@ Vagrant.configure(2) do |config|
   # SHELL
   config.vm.provision "shell", path: "scripts/vagrant_setup.sh"
   config.vm.provision :salt do |salt|
-    salt.bootstrap_options = '-U -Z'
+    salt.bootstrap_options = '-U'
```
Comment (blarghmatey) [inline Vagrantfile]: We should be adding `-x python3` to ensure that we're testing against a Python 3 version of Salt.

File: Vagrantfile [hunk 2]
```diff
@@ -82,11 +95,13 @@ Vagrant.configure(2) do |config|
...
     salt.masterless = true
     salt.run_highstate = true
     salt.colorize = true
     salt.verbose = true
+    salt.install_type = 'git'
+    salt.install_args = 'v2019.2.0'
```
Comment (blarghmatey) [inline Vagrantfile]: We can safely remove this as they have reverted the `latest` package to be 2019.2.0

File: fluentd/map.jinja [hunk 3]
```diff
@@ -11,24 +11,65 @@
             'key_file': 'fluentd.example.com.key'
         },
         'ssl_directory': '/etc/ssl',
-        'nginx_site_path': '/etc/nginx/sites-enabled'
+        'nginx_site_path': '/etc/nginx/sites-enabled',
```
Comment (blarghmatey) [inline fluentd/map.jinja]: We should excise the Nginx functionality from this formula in favor of composing it with the Nginx formula so that we aren't inflating the functionality needlessly.

File: fluentd/map.jinja [hunk 4]
```diff
@@ -11,24 +11,65 @@
...
+            'libffi-devel',
+            'libyaml-devel',
+            'readline-devel',
+            'zlib-devel',
+            'ncurses-devel',
+            'pyOpenSSL'
         ]
     },
-}, grain='os_family', merge=salt.pillar.get('fluentd:overrides'), base='default') %}
+}, grain='os', merge=salt.pillar.get('fluentd:overrides'), base='default') %}
```
Comment (blarghmatey) [inline fluentd/map.jinja]: Why the change to `os` from `os_family`?

File: scripts/vagrant_setup.sh [hunk 5]
```diff
@@ -35,3 +35,6 @@ base:
     - fluentd.plugins
     - fluentd.config
     - fluentd.reverse_proxy" | sudo tee /srv/salt/top.sls
+
+# The salt-minion service needs this:
+sudo sh -c 'echo "127.0.0.1 salt" >> /etc/hosts'
```
Comment (blarghmatey) [inline scripts/vagrant_setup.sh]: You can remove this in favor of adding `-A 127.0.0.1` to the script args in the vagrantfile

File: tests/conftest.py [hunk 6]
```diff
@@ -3,16 +3,16 @@
 
 
 @pytest.fixture()
-def RubyGem(Command):
+def RubyGem(host):
```
Comment (blarghmatey) [inline tests/conftest.py:6]: We should be able to remove this entire file in favor of the rubygem resource built into Testinfra in recent releases.

**Review by blarghmatey — CHANGES_REQUESTED** [review]
For the testinfra tests, we should translate them to use the integration with Salt. The pytest style tests predate that extension and all of the more recent formulas use the salt approach.

File: Vagrantfile (hunk 1 above)
Comment (markbreedlove) [reply Vagrantfile]: Making the 1-line change to use Python 3 produces this error when trying to build a new VM:

```
 *  INFO: Running config_salt()
 *  INFO: Running install_debian_git()
Traceback (most recent call last):
  File "setup.py", line 26, in <module>
    import distutils.dist
ModuleNotFoundError: No module named 'distutils.dist'
 * ERROR: Failed to run install_debian_git()!!!
```

Does this ring a bell at all?
`distutils.dist` is part of the standard library, so I'm not sure what to think yet.

Comment (markbreedlove) [reply Vagrantfile]: I made it use Salt's package by removing `salt.install_type` and `salt.install_args` (see below), but now I get this error:

```
The following packages have unmet dependencies:
 salt-minion : Depends: salt-common (= 2019.2.0+ds-1) but it is not going to be installed
 * ERROR: Failed to run install_debian_stable()!!!
```

... pondering ...

Comment (markbreedlove) [reply Vagrantfile]: It's a known issue with Debian 10. I'll downgrade to Debian 9.

Comment (markbreedlove) [reply Vagrantfile]: I've added `-x python3`.

File: fluentd/map.jinja (hunk 4 above)
Comment (markbreedlove) [reply fluentd/map.jinja]: This was due to there being differences in the names of some development packages between Ubuntu and Debian, but they're not needed anymore, following the decision not to use Rbenv.

File: tests/conftest.py (hunk 6 above)
Comment (markbreedlove) [reply tests/conftest.py:6]: I'm having a hard time finding any documentation on that. Can you point me in the right direction?

Comment (blarghmatey) [reply tests/conftest.py:6]: I was mistaken. There had been some back and forth a while back and I thought that a gem module had been added. We can take a look at upstreaming this later.

**Comment by blarghmatey** [comment]
The only thing left here is to convert the tests to use the testinfra plugin, but we can handle that as a separate issue.

**Review by blarghmatey — APPROVED** [review]

**Comment by blarghmatey** [comment]
:+1:

**Comment by markbreedlove** [comment]
This is ready to merge and release, but I'm holding off until https://github.com/mitodl/salt-ops/pull/539 looks like it's mostly ready to go, in order to coordinate their release & deployment.

**Comment by blarghmatey** [comment]
@markbreedlove this looks ready to merge and shouldn't have any impact on existing infrastructure so there's no need to wait for the salt-ops/pillar piece of things. I'm actually running into a bug due to Salt on Python3 that this will fix so I'm merging now. Thanks!
